src/native/libs/System.Security.Cryptography.Native/pal_evp_pkey_ecdh.c

```diff
@@ -25,43 +23,42 @@ EVP_PKEY_CTX* CryptoNative_EvpPKeyCtxCreate(EVP_PKEY* pkey, EVP_PKEY* peerkey, u
 
     size_t tmpLength = 0;
 
-    /* Initialize, provide the peer public key, and determine the buffer size */
-    if (1 != EVP_PKEY_derive_init(ctx) || 1 != EVP_PKEY_derive_set_peer(ctx, peerkey) ||
-        1 != EVP_PKEY_derive(ctx, NULL, &tmpLength))
+    /* Initialize, provide the peer public key */
+    if (1 != EVP_PKEY_derive_init(ctx) || 1 != EVP_PKEY_derive_set_peer(ctx, peerkey))
     {
         EVP_PKEY_CTX_free(ctx);
         return NULL;
     }
 
-    *secretLength = (uint32_t)tmpLength;
     return ctx;
 }
 
-int32_t CryptoNative_EvpPKeyDeriveSecretAgreement(uint8_t* secret, uint32_t secretLength, EVP_PKEY_CTX* ctx)
+int32_t CryptoNative_EvpPKeyDeriveSecretAgreement(EVP_PKEY* pkey, void* extraHandle, EVP_PKEY* peerKey, uint8_t* secret, uint32_t secretLength)
 {
-    size_t tmpSize = (size_t)secretLength;
-    int ret = 0;
+    if (pkey == NULL || peerKey == NULL || secretLength == 0 || secret == NULL)
+    {
+        return 0;
+    }
 
     ERR_clear_error();
 
-    if (secret != NULL && ctx != NULL)
-    {
-        ret = EVP_PKEY_derive(ctx, secret, &tmpSize);
+    size_t tmpSize = (size_t)secretLength;
+    EVP_PKEY_CTX* ctx = EvpPKeyCtxCreate(pkey, extraHandle, peerKey);
 
-        if (ret == 1 && tmpSize != (size_t)secretLength)
-        {
-            OPENSSL_cleanse(secret, secretLength);
-            ret = 0;
-        }
+    if (ctx == NULL)
+    {
+        return 0;
     }
 
-    return ret;
-}
+    int ret = EVP_PKEY_derive(ctx, secret, &tmpSize);
 
-void CryptoNative_EvpPKeyCtxDestroy(EVP_PKEY_CTX* ctx)
-{
-    if (ctx != NULL)
+    EVP_PKEY_CTX_free(ctx);
+
+    if (ret != 1)
     {
-        EVP_PKEY_CTX_free(ctx);
+        return 0;
     }
+
+    assert(tmpSize > 0 && tmpSize <= secretLength);
```

Comment:
Are asserts enabled in the release(d) builds?  If so, they (I think) tear the process down when they fire.  Is that what we want here?  (If not, are we confident about these values in a release(d) build?)

Reply:
Per docs it's a no-op when NDEBUG is set and cmake does that by default so unless someone overrode that that should be a no-op on release builds.

I'm confident about these values - if OpenSSL updated and somehow started doing OOB writes (which seems extremely unlikely) it's probably better for the process tear down anyway than allow it to continue (OOB write vs DoS) but regardless that's unexpected when they returned `1` - it can really only happen when we refactored and did something wrong like using wrong variable